src/D3MHub.sol

```diff
@@ -0,0 +1,606 @@
+        Once called, subsequent calls to `exec` will unwind as much of the
+        position as possible.
+        @dev msg.sender must be authorized.
+        @param ilk bytes32 of the D3M ilk name
+    */
+    function cage(bytes32 ilk) external auth {
+        require(vat.live() == 1, "D3MHub/no-cage-during-shutdown");
+        require(ilks[ilk].tic == 0, "D3MHub/pool-already-caged");
+
+        ilks[ilk].tic = block.timestamp + ilks[ilk].tau;
+        emit Cage(ilk);
+    }
+
+    /**
+        @notice Write off the debt for a caged pool.
+        This must occur while vat is live. Can be triggered by auth or
+        after tau number of seconds has passed since the pool was caged.
+        @dev This will send the pool's debt to the vow as sin and convert its
+        collateral to gems.  There is a situation where another user has paid
+        back some of the Pool's debt where ink != art in this case we rebalance
+        so that vat.gems(pool) will represent the amount of debt sent to the vow.
+        @param ilk bytes32 of the D3M ilk name
+    */
+    function cull(bytes32 ilk) external {
+        require(vat.live() == 1, "D3MHub/no-cull-during-shutdown");
+
+        uint256 tic = ilks[ilk].tic;
+        require(tic > 0, "D3MHub/pool-live");
+
+        require(tic <= block.timestamp || wards[msg.sender] == 1, "D3MHub/unauthorized-cull");
+
+        uint256 culled = ilks[ilk].culled;
+        require(culled == 0, "D3MHub/already-culled");
+
+        ID3MPool pool = ilks[ilk].pool;
+
+        (uint256 ink, uint256 art) = vat.urns(ilk, address(pool));
+        require(ink <= MAXINT256, "D3MHub/overflow");
+        require(art <= MAXINT256, "D3MHub/overflow");
+        vat.grab(ilk, address(pool), address(pool), vow, -int256(ink), -int256(art));
+
+        if (ink > art) {
+            // We have more collateral than debt, so need to rebalance.
+            // After cull the gems we grab above represent the debt to
+            // unwind.
+            unchecked {
+                vat.slip(ilk, address(pool), -int256(ink - art));
+            }
+        }
+
+        ilks[ilk].culled = 1;
+        emit Cull(ilk, ink, art);
+    }
+
+    /**
+        @notice Rollback Write-off (cull) if General Shutdown happened.
+        This function is required to have the collateral back in the vault so it
+        can be taken by End module and eventually be shared to DAI holders (as
+        any other collateral) or maybe even unwinded.
+        @dev This pulls gems from the pool and reopens the urn with the gem
+        amount of ink/art.
+        @param ilk bytes32 of the D3M ilk name
+    */
+    function uncull(bytes32 ilk) external {
+        ID3MPool pool = ilks[ilk].pool;
+
+        require(ilks[ilk].culled == 1, "D3MHub/not-prev-culled");
+        require(vat.live() == 0, "D3MHub/no-uncull-normal-operation");
+
+        address vow_ = vow;
+        uint256 wad = vat.gem(ilk, address(pool));
+        require(wad <= MAXINT256, "D3MHub/overflow");
```

Comment:
Review this case of overflow.

Reply:
I believe this solves this issue: https://github.com/makerdao/dss-direct-deposit/pull/46/commits/c3b9db9890ebcbf7e27773319dd6ac6014a02b0d